telemetry/otel.go

```diff
@@ -0,0 +1,234 @@
+	tracerProviderFunc func(context.Context, *resource.Resource) (*trace.TracerProvider, error)
+)
+
+// Start initializes automatic instrumentation for exporting OTLP metrics and traces
+// to an OpenTelemetry collector agent.
+func Start(ctx context.Context) (shutdown func(context.Context) error, err error) {
+	defaultStartOnce.Do(func() {
+		shutdown, err = start(ctx, newOTLPTraceProvider, newOTLPMeterProvider)
+	})
+	return
+}
+
+// StartWithStdoutExport initializes automatic instrumentation for exporting OTLP
+// metrics and traces to stdout (e.g. Cloud Logging if using Cloud Run).
+//
+// This is intended for testing and debugging metrics and traces before exporting
+// them to an OpenTelemetry collector agent.
+func StartWithStdoutExport(ctx context.Context) (shutdown func(context.Context) error, err error) {
+	defaultStartOnce.Do(func() {
+		shutdown, err = start(ctx, newStdoutTraceProvider, newStdoutMeterProvider)
+	})
+	return
+}
+
+func start(ctx context.Context, newTracer tracerProviderFunc, newMeter meterProviderFunc) (shutdown func(context.Context) error, err error) {
+	var shutdownFuncs []func(context.Context) error
+
+	shutdown = func(ctx context.Context) error {
+		var err error
+		for _, fn := range shutdownFuncs {
+			err = errors.Join(err, fn(ctx))
+		}
+		shutdownFuncs = nil
+		return err
+	}
+
+	if err := runtime.Start(); err != nil {
+		return nil, fmt.Errorf("failed to start runtime instrumentation: %w", err)
+	}
+	// Set up some automatic attributes regarding deployment environment, service,
+	// GCP, etc to attach to telemetry.
+	res, err := resource.New(
+		ctx,
+		resource.WithFromEnv(),
+		resource.WithProcess(),
+		resource.WithDetectors(gcp.NewDetector()),
+		resource.WithTelemetrySDK(),
+		resource.WithHost(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create opentelemetry resource: %w", err)
+	}
+
+	// Set up propagator to W3C trace context, a common specification for
+	// distributed tracing. See https://opentelemetry.io/docs/concepts/context-propagation.
+	prop := newPropagator()
+	otel.SetTextMapPropagator(prop)
+
+	// Set up tracing.
+	tracerProvider, err := newTracer(ctx, res)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set up tracing provider: %w", err)
+	}
+	shutdownFuncs = append(shutdownFuncs, tracerProvider.Shutdown)
+	otel.SetTracerProvider(tracerProvider)
+
+	// Set up monitoring.
+	meterProvider, err := newMeter(ctx, res)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set up meter provider: %w", err)
+	}
+	shutdownFuncs = append(shutdownFuncs, meterProvider.Shutdown)
+	otel.SetMeterProvider(meterProvider)
+
+	return
+}
+
+func newPropagator() propagation.TextMapPropagator {
+	return propagation.NewCompositeTextMapPropagator(
+		propagation.TraceContext{},
+		propagation.Baggage{},
+	)
+}
+
+func newOTLPTraceProvider(ctx context.Context, res *resource.Resource) (*trace.TracerProvider, error) {
+	otelAgentAddr, ok := os.LookupEnv("OTEL_EXPORTER_OTLP_ENDPOINT")
```

Comment:
Should we be using https://github.com/abcxyz/pkg/tree/main/cfgloader ?

Reply:
I took a look. How would it be more useful here? Adding a config with only OTEL_EXPORTER_OTLP_ENDPOINT seems overkill at this point. 

```
// Package cfgloader provides common functionality to load configs.
// It's meant for loading configs from YAML and/or env vars. Don't use
// it for initializing configs with defaults.
```